File: artifacts/ai-business-os/src/services/ai-workforce.ts
import { apiClient, type ApiClient } from "./api-client.ts";

export type AgentRole =
  | "business_manager"
  | "cmo"
  | "sales"
  | "support"
  | "operations"
  | "analytics";
export type AutonomyMode = "manual" | "supervised" | "autonomous";

export type AgentCapability = {
  key: string;
  category: "read" | "analysis" | "draft" | "governed_action";
  description: string;
};

export type AgentMetrics = {
  execution_count: number;
  completed_count: number;
  needs_approval_count: number;
  failed_count: number;
  average_duration_ms: number | null;
  proposed_action_count: number;
  pending_approval_count: number;
  approval_rate: number | null;
  input_tokens: number;
  output_tokens: number;
};

export type AgentConfig = {
  id: string;
  business_id: string;
  role: AgentRole;
  display_name: string;
  enabled: boolean;
  status: "active" | "disabled";
  health: "ready" | "not_configured";
  autonomy_mode: AutonomyMode;
  autonomy_description: string;
  custom_instructions: string | null;
  capabilities: AgentCapability[];
  default_capabilities: string[];
  role_description: string;
  metrics: AgentMetrics;
  last_activity_at: string | null;
  created_at: string;
  updated_at: string;
};

export type ApprovalLink = { id: string; status: string; reason_code: string };
export type AgentProposedAction = {
  id: string;
  execution_id: string;
  action_type: string;
  description: string;
  risk_level: string;
  status: string;
  policy_decision: string | null;
  requires_approval: boolean;
  approval: ApprovalLink | null;
};

export type AgentActivity = {
  id: string;
  business_id: string;
  command_id: string | null;
  parent_execution_id: string | null;
  role: AgentRole;
  trigger: string;
  status: string;
  task_summary: string;
  summary: string | null;
  failure_code: string | null;
  duration_ms: number | null;
  input_tokens: number | null;
  output_tokens: number | null;
  estimated_cost_usd: string | null;
  delegation_sequence: number;
  delegation_depth: number;
  proposed_actions: AgentProposedAction[];
  created_at: string;
  completed_at: string | null;
};

export type PageResponse<T> = { items: T[]; page: number; page_size: number; total: number };
export type CommandRoute = {
  primary_role: AgentRole;
  intent: string;
  required_capabilities: string[];
  relevant_modules: string[];
  delegation_roles: AgentRole[];
  clarification_required: boolean;
};
export type AICommand = {
  id: string;
  business_id: string;
  requested_by_user_id: string | null;
  command: string;
  status: "queued" | "running" | "completed" | "needs_approval" | "failed" | "canceled";
  route: CommandRoute;
  execution_id: string | null;
  summary: string | null;
  failure_code: string | null;
  executions: AgentActivity[];
  proposed_actions: AgentProposedAction[];
  created_at: string;
  completed_at: string | null;
};
export type SuggestedCommand = { command: string; reason: string; role: AgentRole };
export type DailyBrief = {
  generated_at: string;
  sections: Array<{ key: string; title: string; facts: string[] }>;
  recommended_priorities: string[];
};
export type CommandContextReference = {
  type: "customer" | "lead" | "order" | "conversation" | "appointment_type" | "provider" | "campaign" | "workflow" | "report";
  id: string;
};

const path = (businessId: string, suffix: string) =>
  `/api/v1/businesses/${encodeURIComponent(businessId)}${suffix}`;

export function createAIWorkforceApi(client: ApiClient) {
  return {
    agents: {
      list: (businessId: string, signal?: AbortSignal) =>
        client.request<AgentConfig[]>(path(businessId, "/agents"), { signal }),
      get: (businessId: string, role: AgentRole, signal?: AbortSignal) =>
        client.request<AgentConfig>(path(businessId, `/agents/${role}`), { signal }),
      capabilities: (businessId: string, signal?: AbortSignal) =>
        client.request<AgentCapability[]>(path(businessId, "/agents/capabilities"), { signal }),
      update: (
        businessId: string,
        role: AgentRole,
        data: Partial<Pick<AgentConfig, "display_name" | "enabled" | "autonomy_mode">> & {
          custom_instructions?: string | null;
          capabilities?: string[];
        },
      ) => client.request<AgentConfig>(path(businessId, `/agents/${role}`), { method: "PATCH", json: data }),
      reset: (businessId: string, role: AgentRole) =>
        client.request<AgentConfig>(path(businessId, `/agents/${role}/reset`), { method: "POST" }),
      activity: (
        businessId: string,
        options: { page?: number; pageSize?: number; role?: AgentRole; status?: string } = {},
        signal?: AbortSignal,
      ) => {
        const query = new URLSearchParams({
          page: String(options.page ?? 1),
          page_size: String(options.pageSize ?? 25),
        });
        if (options.role) query.set("role", options.role);
        if (options.status) query.set("status", options.status);
        return client.request<PageResponse<AgentActivity>>(
          path(businessId, `/agents/activity?${query}`),
          { signal },
        );
      },
      activityDetail: (businessId: string, executionId: string, signal?: AbortSignal) =>
        client.request<AgentActivity>(path(businessId, `/agents/activity/${executionId}`), { signal }),
    },
    commands: {
      execute: (
        businessId: string,
        command: string,
        trigger_source: "command_center" | "dashboard" | "agent_detail" = "command_center",
        context_references: CommandContextReference[] = [],
      ) => client.request<AICommand>(path(businessId, "/commands"), {
        method: "POST",
        json: { command, trigger_source, context_references },
      }),
      list: (businessId: string, page = 1, signal?: AbortSignal) =>
        client.request<PageResponse<AICommand>>(path(businessId, `/commands?page=${page}&page_size=20`), { signal }),
      get: (businessId: string, commandId: string, signal?: AbortSignal) =>
        client.request<AICommand>(path(businessId, `/commands/${commandId}`), { signal }),
      cancel: (businessId: string, commandId: string) =>
        client.request<AICommand>(path(businessId, `/commands/${commandId}/cancel`), { method: "POST" }),
      suggestions: (businessId: string, signal?: AbortSignal) =>
        client.request<SuggestedCommand[]>(path(businessId, "/commands/suggestions"), { signal }),
      dailyBrief: (businessId: string, signal?: AbortSignal) =>
        client.request<DailyBrief>(path(businessId, "/commands/daily-brief"), { signal }),
    },
  };
}

export const aiWorkforceApi = createAIWorkforceApi(apiClient);

File: backend/app/schemas/ai_workforce.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.domain.ai_workforce import CommandIntent
from app.schemas.ai_agent import AIAgentRole
from app.services.ai_capabilities import CapabilityCategory


class CapabilityResponse(BaseModel):
    key: str
    category: CapabilityCategory
    description: str
    model_config = ConfigDict(extra="forbid")


class AgentMetricsResponse(BaseModel):
    execution_count: int = 0
    completed_count: int = 0
    needs_approval_count: int = 0
    failed_count: int = 0
    average_duration_ms: int | None = None
    proposed_action_count: int = 0
    pending_approval_count: int = 0
    approval_rate: float | None = None
    input_tokens: int = 0
    output_tokens: int = 0
    model_config = ConfigDict(extra="forbid")


class AgentConfigResponse(BaseModel):
    id: UUID
    business_id: UUID
    role: AIAgentRole
    display_name: str
    enabled: bool
    status: Literal["active", "disabled"]
    health: Literal["ready", "not_configured"]
    autonomy_mode: Literal["manual", "supervised", "autonomous"]
    autonomy_description: str
    custom_instructions: str | None
    capabilities: list[CapabilityResponse]
    default_capabilities: list[str]
    role_description: str
    metrics: AgentMetricsResponse = Field(default_factory=AgentMetricsResponse)
    last_activity_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(extra="forbid")


class AgentConfigUpdate(BaseModel):
    display_name: str | None = Field(default=None, min_length=1, max_length=100)
    enabled: bool | None = None
    autonomy_mode: Literal["manual", "supervised", "autonomous"] | None = None
    custom_instructions: str | None = Field(default=None, max_length=2_000)
    capabilities: list[str] | None = Field(default=None, max_length=40)
    model_config = ConfigDict(extra="forbid")

    @field_validator("display_name", "custom_instructions", mode="before")
    @classmethod
    def trim_text(cls, value: Any) -> Any:
        if isinstance(value, str):
            value = value.strip()
        return value or None

    @field_validator("capabilities")
    @classmethod
    def validate_capabilities(cls, value: list[str] | None) -> list[str] | None:
        if value is None:
            return None
        normalized = [item.strip() for item in value if isinstance(item, str)]
        if len(normalized) != len(value) or any(not item for item in normalized):
            raise ValueError("Capabilities must contain non-blank strings")
        if len(set(normalized)) != len(normalized):
            raise ValueError("Capabilities cannot contain duplicates")
        return normalized

    @model_validator(mode="after")
    def require_change(self) -> "AgentConfigUpdate":
        if not self.model_fields_set:
            raise ValueError("At least one configuration field is required")
        return self


class CommandContextReference(BaseModel):
    type: Literal[
        "customer", "lead", "order", "conversation", "appointment_type",
        "provider", "campaign", "workflow", "report", "integration_connection",
    ]
    id: UUID
    model_config = ConfigDict(extra="forbid")


class CommandCreateRequest(BaseModel):
    command: str = Field(min_length=1, max_length=4_000)
    context_references: list[CommandContextReference] = Field(default_factory=list, max_length=10)
    trigger_source: Literal["command_center", "dashboard", "agent_detail"] = "command_center"
    model_config = ConfigDict(extra="forbid")

    @field_validator("command", mode="before")
    @classmethod
    def trim_command(cls, value: Any) -> Any:
        return value.strip() if isinstance(value, str) else value

    @field_validator("context_references")
    @classmethod
    def unique_references(cls, value: list[CommandContextReference]) -> list[CommandContextReference]:
        keys = [(item.type, item.id) for item in value]
        if len(keys) != len(set(keys)):
            raise ValueError("Context references cannot contain duplicates")
        return value


class CommandRouteResponse(BaseModel):
    primary_role: AIAgentRole
    intent: CommandIntent
    required_capabilities: list[str]
    relevant_modules: list[str]
    delegation_roles: list[AIAgentRole]
    clarification_required: bool = False
    model_config = ConfigDict(extra="forbid")


class ApprovalLinkResponse(BaseModel):
    id: UUID
    status: str
    reason_code: str
    model_config = ConfigDict(extra="forbid")


class ProposedActionResponse(BaseModel):
    id: UUID
    execution_id: UUID
    action_type: str
    description: str
    risk_level: str
    status: str
    policy_decision: str | None
    requires_approval: bool
    approval: ApprovalLinkResponse | None = None
    model_config = ConfigDict(extra="forbid")


class AgentActivityResponse(BaseModel):
    id: UUID
    business_id: UUID
    command_id: UUID | None
    parent_execution_id: UUID | None
    role: AIAgentRole
    trigger: str
    status: str
    task_summary: str
    summary: str | None
    failure_code: str | None
    duration_ms: int | None
    input_tokens: int | None
    output_tokens: int | None
    estimated_cost_usd: Decimal | None
    delegation_sequence: int
    delegation_depth: int
    proposed_actions: list[ProposedActionResponse]
    created_at: datetime
    completed_at: datetime | None
    model_config = ConfigDict(extra="forbid")


class AgentActivityPage(BaseModel):
    items: list[AgentActivityResponse]
    page: int
    page_size: int
    total: int
    model_config = ConfigDict(extra="forbid")


class CommandResponse(BaseModel):
    id: UUID
    business_id: UUID
    requested_by_user_id: UUID | None
    command: str
    status: str
    route: CommandRouteResponse
    execution_id: UUID | None
    summary: str | None
    failure_code: str | None
    executions: list[AgentActivityResponse] = Field(default_factory=list)
    proposed_actions: list[ProposedActionResponse] = Field(default_factory=list)
    created_at: datetime
    completed_at: datetime | None
    model_config = ConfigDict(extra="forbid")


class CommandPage(BaseModel):
    items: list[CommandResponse]
    page: int
    page_size: int
    total: int
    model_config = ConfigDict(extra="forbid")


class SuggestedCommandResponse(BaseModel):
    command: str
    reason: str
    role: AIAgentRole
    model_config = ConfigDict(extra="forbid")


class DailyBriefSection(BaseModel):
    key: str
    title: str
    facts: list[str]
    model_config = ConfigDict(extra="forbid")


class DailyBriefResponse(BaseModel):
    generated_at: datetime
    sections: list[DailyBriefSection]
    recommended_priorities: list[str]
    model_config = ConfigDict(extra="forbid")
